Do not attempt to send email when no "From Email" set

PhilETaylor opened the pull request:
Pull Request for Issue #16093 .

### Summary of Changes

Dont error if there is no from EMAIL in Joomla Global Config, when sending an email.

### Testing Instructions

install Joomla 3.7.1
REMOVE the from name in your global configuration mail settings
Create a new Publisher using the admin console.

### Expected result

No email is sent, no PHP warning given.

### Actual result

> 0 Call to a member function addRecipient() on boolean

### After applying this patch

<img width="577" alt="screen shot 2017-05-18 at 13 09 33" src="https://cloud.githubusercontent.com/assets/400092/26201564/87b2242a-3bcb-11e7-8a4a-c7d07d54f97b.png">


### Documentation Changes Required

(Note: PR quickly created in Github, may need refining)

Commit:
Update joomla.php

mbabker commented:
Still doesn't fully fix it.  If PHPMailer rejects the from or recipient addresses and returns a boolean false, it'll break the fluent interface.

PhilETaylor commented:
It fixes the reported issue, which is where there is no Mail From email address set in Joomla Global Config. 

It doesn't attempt to fix the issues you are reporting ;-)

PhilETaylor commented:
Are you advocating removing the "sexy way" of chaining the method calls when using phpMailer then?

mbabker commented:
It's twofold.  You fix an obvious "if this field doesn't have an `@` symbol then it's totally broken and can't work" issue.  But if the supplied email address values fail PHPMailer's validation when adding addresses, then it will still break because `setSender` and `addRecipient` can both return either `$this` or boolean false, the latter breaking the fluent interface.

So ya, you'd have to drop the "sexy" method chaining here.

PhilETaylor commented:
ok thats an "in the office job" and not "in the hotel lobby drinking coffee" job :)

franz-wohlkoenig commented:
I have tested this item :white_check_mark: successfully on 4f1e902fe977fc9705eb87c3b5eea3582f2e5143<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/16107">issues.joomla.org/tracker/joomla-cms/16107</a>.</sub>

PhilETaylor commented:
No point testing @franz-wohlkoenig Im gonna improve this when I get to the office.

Commit:
Remove fluent interface, should not be used in this instance

Commit:
Remove fluent interface, should not be used in this instance

PhilETaylor commented:
I have now removed the fluent interface use of phpMailer calls, please test again.

PhilETaylor commented:
@Flowman @joo7 @sharphosting Please test this PR as you reported in #16093

franz-wohlkoenig commented:
@PhilETaylor got different Info than before:
![bildschirmfoto 2017-05-20 um 12 25 39](https://cloud.githubusercontent.com/assets/8235763/26275019/81d06a5c-3d57-11e7-9205-138a6b74ac64.JPG)
if this is correct, i set "sucessfully Test".

brianteeman commented:
I can confirm the issue before the pr
but after applying the pr I just get the message user successfully saved - no warning or error
![screen shot 2017-05-22 at 18 16 29](https://issues.joomla.org/uploads/1/0ee4a1afcb42cdb63f23c8b6979fdb27.png)<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/16107">issues.joomla.org/tracker/joomla-cms/16107</a>.</sub>

brianteeman commented:
I have tested this item :red_circle: unsuccessfully on d861f6c0264fa8519a4cf9955843956e4735e905<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/16107">issues.joomla.org/tracker/joomla-cms/16107</a>.</sub>

PhilETaylor commented:
thats the joy with php mail, it provides no feedback…


> On 22 May 2017, at 19:16, Brian Teeman <notifications@github.com> wrote:
> 
> I have tested this item 🔴 unsuccessfully on d861f6c <https://github.com/joomla/joomla-cms/commit/d861f6c0264fa8519a4cf9955843956e4735e905>
> This comment was created with the J!Tracker Application <https://github.com/joomla/jissues> at issues.joomla.org/tracker/joomla-cms/16107 <https://issues.joomla.org/tracker/joomla-cms/16107>.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/joomla/joomla-cms/pull/16107#issuecomment-303179036>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AAYa3HChRUrciKv6YLnGzvJKgqOZZjJ4ks5r8dEOgaJpZM4NfIBi>.
>

brianteeman commented:
i think i would prefer the ugly error message then from before the pr in that case. at least i know something has gone wrong then<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/16107">issues.joomla.org/tracker/joomla-cms/16107</a>.</sub>

mbabker commented:
You can check if `setSender` and `addRecipient` return a boolean false instead of ignoring the return in those cases.  In theory, https://github.com/joomla/joomla-cms/pull/16107/files#diff-8b42e55d5e424b7f172d227f1a012e1dR167 should be reporting a sending failure but it might have to become a `if ($mail->Send() !== true)` check (because IIRC that method can return boolean true, boolean false, or `JException`).

Commit:
Belt and braces approach, remove nested if statements, cleanup

PhilETaylor commented:
I have now reworked this PR, cleaned up some nested if statements and implemented the suggestions. 

Testing instructions:

1) Go to Joomla Global Config, remove your From Email Address, save config
2) Go to User Manager, create a NEW user with level "Publisher"
3) Save and Close this user - see the message 

<img width="461" alt="screen shot 2017-05-23 at 10 31 45" src="https://cloud.githubusercontent.com/assets/400092/26347934/0a6c4afa-3fa3-11e7-825f-0a62fd20e8c3.png">

franz-wohlkoenig commented:
I have tested this item :white_check_mark: successfully on 64cf52ad010a842d85234412f5ca172f8933b46b<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/16107">issues.joomla.org/tracker/joomla-cms/16107</a>.</sub>

Quy commented on plugins/user/joomla/joomla.php:
```diff
@@ -101,79 +101,85 @@ public function onUserAfterSave($user, $isnew, $success, $msg)
 		$mail_to_user = $this->params->get('mail_to_user', 1);
 
-		if ($isnew)
+		if (!$isnew || !$mail_to_user) {
```
`{` on its own line.

Quy commented on plugins/user/joomla/joomla.php:
```diff
@@ -101,79 +101,85 @@ public function onUserAfterSave($user, $isnew, $success, $msg)
+
+		// TODO: Suck in the frontend registration emails here as well. Job for a rainy day.
+		// The method check here ensures that if running as a CLI Application we don't get any errors
+		if (method_exists($this->app, 'isClient') && !$this->app->isClient('administrator')) {
```
`{` on its own line.

Quy commented on plugins/user/joomla/joomla.php:
```diff
@@ -101,79 +101,85 @@ public function onUserAfterSave($user, $isnew, $success, $msg)
+			$lang->setLanguage($defaultLocale);
+		}
+
+		if ($mail->Send() !== true) {
```
`{` on its own line.

Quy commented on plugins/user/joomla/joomla.php:
```diff
@@ -101,79 +101,85 @@ public function onUserAfterSave($user, $isnew, $success, $msg)
+		$mail->setBody($emailBody);
+
+		// Set application language back to default if we changed it
+		if ($userLocale != $defaultLocale) {
```
`{` on its own line.

Commit:
Codestyle

PhilETaylor commented:
@quy Ive reset those braces, darn IDE :)

Quy commented on plugins/user/joomla/joomla.php:
```diff
@@ -101,79 +101,85 @@ public function onUserAfterSave($user, $isnew, $success, $msg)
+
+		$retSetRecipient = $mail->addRecipient($user['email']);
+
+		if (!$retSetSender || $retSetRecipient)
```
I'm not at a computer to test. Should `$retSetRecipient''` be `!$retSetRecipient`?

Commit:
Missing !

Quy commented on plugins/user/joomla/joomla.php:
```diff
@@ -101,79 +101,85 @@ public function onUserAfterSave($user, $isnew, $success, $msg)
+		}
+
+		// check if we have a sensible from email address, if not bail out as mail would not be sent anyway
+		if (strpos($this->app->get('mailfrom'), '@'))
```
It should be `if (strpos($this->app->get('mailfrom'), '@') === false)` to find no @.

Quy commented:
There is a similar pr #13518, Maybe it will you some insights to incorporate techniques if any such as `try/catch` and `send()`.

Commit:
check for a @ as a sanity check

Commit:
revert all changes and use sendMail instead.

PhilETaylor commented:
I have changed this PR so that we use ` JFactory::getMailer()->sendMail()` function instead, this handles all the if statements already and so no need to repeat them, just consume the API.

Commit:
Codestyle

franz-wohlkoenig commented:
@PhilETaylor need to test again?

PhilETaylor commented:
Yes Please.

brianteeman commented:
yes you need to test again as the code has changed

franz-wohlkoenig commented:
I have tested this item :white_check_mark: successfully on 3c49f1f72080a6ecc32419aa7575ccbf37b03c0f<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/16107">issues.joomla.org/tracker/joomla-cms/16107</a>.</sub>

Quy commented:
I have tested this item :white_check_mark: successfully on 3c49f1f72080a6ecc32419aa7575ccbf37b03c0f<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/16107">issues.joomla.org/tracker/joomla-cms/16107</a>.</sub>

franz-wohlkoenig commented:
RTC after two successful tests.
